File: cuboid.py
import math

class GeometryLibrary:
    pass

class Point:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z


    def __eq__(self, other):
        return self.x == other.x and self.y == other.y and self.z == other.z


class Cuboid(GeometryLibrary):
    def __init__(self, vertices):
        if len(vertices) != 8:
            raise ValueError("A cuboid must have only 8 vertices")
        self.vertices = vertices


    def __eq__(self, other):
        return all(v1 == v2 for v1, v2 in zip(self.vertices, other.vertices))

    
    def centroid(self):
        x = sum(vertex.x for vertex in self.vertices) / 8
        y = sum(vertex.y for vertex in self.vertices) / 8
        z = sum(vertex.z for vertex in self.vertices) / 8
        return Point(x, y, z)


    def volume(self):
        d1 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:4], self.vertices[4:])))
        d2 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:2] + self.vertices[6:8], self.vertices[2:6])))
        d3 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:1] + self.vertices[3:5] + self.vertices[6:], self.vertices[1:3] + self.vertices[5:6])))
        return d1 * d2 * d3

    def surface_area(self):
        d1 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:4], self.vertices[4:])))
        d2 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:2] + self.vertices[6:8], self.vertices[2:6])))
        d3 = math.sqrt(sum(((v1.x - v2.x) ** 2 + (v1.y - v2.y) ** 2 + (v1.z - v2.z) ** 2) for v1, v2 in zip(self.vertices[:1] + self.vertices[3:5] + self.vertices[6:], self.vertices[1:3] + self.vertices[5:6])))
        return 2 * (d1 * d2 + d1 * d3 + d2 * d3)
    


def main():
    # define the vertices of the cuboid
    vertices = [
        Point(0, 0, 0),
        Point(1, 0, 0),
        Point(1, 1, 0),
        Point(0, 1, 0),
        Point(0, 0, 1),
        Point(1, 0, 1),
        Point(1, 1, 1),
        Point(0, 1, 1)
    ]

    # create a cuboid object
    cuboid = Cuboid(vertices)
    # print(cuboid)

    # calculate the print the centroid, volume and surface area of the cuboid
    centroid = cuboid.centroid()
    volume = cuboid.volume()
    surface_area = cuboid.surface_area()


    print(f"Centroid: ({centroid.x} {centroid.y} {centroid.z})")
    print(f"Volume: {volume}")
    print(f"Surface Area: {surface_area}")




    


if __name__ == "__main__":
    main()

File: cuboid.cs
using System;
using System.Linq;
using GeometryLibrary;

namespace GeometryLibrary
{
    public struct Point3D
    {
        public double X { get; set;}
        public double Y { get; set;}
        public double Z { get; set;}

        public Point3D(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }
    }

    public class Cuboid
    {
        private Point3D[] vertices;

        public Cuboid(Point3D[] vertices)
        {
            if (vertices.Length != 8)
            {
                throw new ArgumentException("A cuboid must have 8 vertices");
            }

            this.vertices = vertices;
        }

        public static bool operator == (Cuboid a, Cuboid b)
        {
            if (ReferenceEquals(a, null))
            {
                return ReferenceEquals(b, null);
            }

            if (ReferenceEquals(b, null))
            {
                return false;
            }

            // assuming two cuboids are equal if they have the same vertices in any order
            return a.vertices.All(v1 => b.vertices.Any(v2 => v1.Equals(v2))) && 
                b.vertices.All(v1=> a.vertices.Any(v2 => v1.Equals(v2)));
        }


        public static bool operator !=(Cuboid a, Cuboid b)
        {
            return !(a == b);
        }

        public Point3D Centroid()
        {
            double x = 0;
            double y = 0;
            double z = 0;

            for (int i = 0; i < 8; i++)
            {
                x += vertices[i].X;
                y += vertices[i].Y;
                z += vertices[i].Z;
            }

            return new Point3D(x /8, y/ 8, z /8);
        }

        public double Volume()
        {
            double width = Math.Abs(vertices[1].X - vertices[0].X);
            double height = Math.Abs(vertices[3].Y - vertices[0].Y);
            double depth = Math.Abs(vertices[4].Z - vertices[0].Z);
            
            return width * depth * height; 
        }

        public double SurfaceArea()
        {
            double width = Math.Abs(vertices[1].X - vertices[0].X);
            double height = Math.Abs(vertices[3].Y - vertices[0].Y);
            double depth = Math.Abs(vertices[4].Z - vertices[0].Z);

            return 2 * (width * height + width * depth + height * depth);
        }

        // 

        public override bool Equals(object obj)
        {
            if(obj == null || GetType() != obj.GetType())
            {
                return false;
            }

            return this == (Cuboid)obj;
        }

        // 
        public override int GetHashCode()
        {
            int hashCode = 0;
            for(int i = 0; i < 8; i++)
            {
                hashCode ^= vertices[i].GetHashCode();
            }
            return hashCode;

        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        Point3D p1 = new Point3D(0, 0, 0);
        Point3D p2 = new Point3D(1, 0, 0);
        Point3D p3 = new Point3D(1, 1, 0);
        Point3D p4 = new Point3D(0, 1, 0);
        Point3D p5 = new Point3D(0, 0, 1);
        Point3D p6 = new Point3D(1, 0, 1);
        Point3D p7 = new Point3D(1, 1, 1);
        Point3D p8 = new Point3D(0, 1, 1);

        Cuboid cuboid = new Cuboid(new Point3D[] { p1, p2, p3, p4, p5, p6, p7, p8});

        Point3D centroid = cuboid.Centroid();
        double Volume = cuboid.Volume();
        double SurfaceArea = cuboid.SurfaceArea();

        Console.WriteLine($"Centroid: ({centroid.X}, {centroid.Y}, {centroid.Z})");
        Console.WriteLine($"Volume: {Volume}");
        Console.WriteLine($"Surface Area: {SurfaceArea}");
    }
}